man/spi-pipe.1
.TH SPI-PIPE "1" "August 2014" "SPI-tools" "User Commands"
.SH NAME
spi-pipe \- full-duplex SPI communication tool.
.SH SYNOPSIS
.B spi-pipe
\fIoptions\fR...
.SH DESCRIPTION
.PP
This program allows full-duplex communications, sending data from its
standard input to the SPI slave, and displaying on its standard output the
data received from the SPI port.
.TP
\fB\-d\fR, \fB\-\-device\fR=\fIDEVICE\fR
use the given Linux spidev character device.
.TP
\fB\-s\fR, \fB\-\-speed\fR=\fI<int>\fR
set the target SPI speed for transfer.
.TP
\fB\-b\fR, \fB\-\-blocksize\fR=\fI<int>\fR
set the block size (in bytes) for transfer.
.TP
\fB\-n\fR, \fB\-\-number\fR=\fI<int>\fR
set the number of blocks to transmit (-1 for continuous transfert).
.TP
\fB\-h\fR, \fB\-\-help\fR
display the help screen and exit.
.TP
\fB\-v\fR, \fB\-\-version\fR
display the version number and exit.

.SH EXAMPLES
.TP
Get help:
$ \fBspi\-pipe \-h\fR
.TP
Send and receive at the same time
$ \fIcommand-1\fR | \fBspi\-pipe \-\-device=/dev/spidev0.0 \fR | \fIcommand-2\fR
.br
Note that \fIcommand-1\fR, \fIcommand-2\fR and \fBspi\-pipe\fR run simultaneously
in three different processes.
.TP
Send data to the SPI link
$ \fIcommand-1\fR | \fBspi\-pipe \-\-device=/dev/spidev0.0 \fR
.TP
Receive data from the SPI link
$ \fBspi\-pipe \-\-device=/dev/spidev0.0 < /dev/zero\fR | \fIcommand-2\fR
.br
You can also use \fIcommand-2 < /dev/spidev0.0\fR but with \fBspi-pipe\fR you can control
what is sent to the device (always `0` here).
.TP
Read 40 blocks of 4 bytes from the SPI link
$  \fBspi\-pipe \-\-device=/dev/spidev0.0 \-b 4 \-n 40 < /dev/zero\fR | \fIcommand-2\fR
.SH AUTHOR
Written by Christophe BLAESS <https://www.blaess.fr/christophe>.
.SH "REPORTING BUGS"
.br
Github home page: <https://github.com/cpb-/spi-tools.git>
.SH COPYRIGHT
Copyright \(co 2014 Christophe Blaess.
.br
Licensed under GPLv2 <http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
.br
There is NO WARRANTY, to the extent permitted by law.


src/spi-pipe.c
/*
 * spidev data transfer tool.
 *
 * (c) 2014-2021 Christophe BLAESS <christophe.blaess@logilin.fr>
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms and conditions of the GNU General Public License,
 * version 2, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St - Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */

#include <fcntl.h>
#include <getopt.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include "config.h"
#include "spi-tools.h"

static char *project = "spi-pipe";

static void display_usage(const char *name)
{
	fprintf(stderr, "usage: %s options...\n", name);
	fprintf(stderr, "  options:\n");
	fprintf(stderr, "    -d --device=<dev>    Use the given spi-dev character device.\n");
	fprintf(stderr, "    -m --mode=[0-3]      Use the selected spi mode:\n");
	fprintf(stderr, "             0: low idle level, sample on leading edge,\n");
	fprintf(stderr, "             1: low idle level, sample on trailing edge,\n");
	fprintf(stderr, "             2: high idle level, sample on leading edge,\n");
	fprintf(stderr, "             3: high idle level, sample on trailing edge.\n");
	fprintf(stderr, "    -s --speed=<speed>   Maximum SPI clock rate (in Hz).\n");
	fprintf(stderr, "    -l --lsb={0,1}     LSB first (1) or MSB first (0).\n");
	fprintf(stderr, "    -B --bits=[7...]     Bits per word.\n");
	fprintf(stderr, "    -b --blocksize=<int> transfer block size in byte.\n");
	fprintf(stderr, "    -n --number=<int>    number of blocks to transfer (-1 = infinite).\n");
	fprintf(stderr, "    -h --help            this screen.\n");
	fprintf(stderr, "    -v --version         display the version number.\n");
}




int main (int argc, char *argv[])
{
	int opt;
	int long_index = 0;

	static struct option options[] = {
		{"device",    required_argument, NULL,  'd' },
		{"speed",     required_argument, NULL,  's' },
		{"blocksize", required_argument, NULL,  'b' },
		{"number",    required_argument, NULL,  'n' },
		{"lsb",       required_argument, NULL,  'l' },
		{"mode",      required_argument, NULL,  'm' },
		{"bits",      required_argument, NULL,  'B' },
		{"help",      no_argument,       NULL,  'h' },
		{"version",   no_argument,       NULL,  'v' },
		{0,           0,                 0,  0 }
	};

	int fd;
	char *device = NULL;
	uint8_t *rx_buffer = NULL;
	uint8_t *tx_buffer = NULL;
	spi_config_t prev_config;
	spi_config_t new_config;

	int block_size    =  1;
	int blocks_count  = -1;
	int offset        =  0;
	int nb            =  0;

	new_config.spi_mode = -1;
	new_config.lsb_first = -1;
	new_config.bits_per_word = -1;
	new_config.spi_speed = -1;
	new_config.spi_ready = -1;


	while ((opt = getopt_long(argc, argv, "d:s:b:l:m:n:B:rhv", options, &long_index)) >= 0) {
		switch(opt) {
			case 'h':
				display_usage(argv[0]);
				exit(EXIT_SUCCESS);

			case 'v':
				fprintf(stderr, "%s - %s\n", project, VERSION);
				fprintf(stderr, "Copyright (c) 2014-2021 Christophe Blaess. (license GPLv2)\n");
				fprintf(stderr, "This is free software. You are free to change and redistribute it.\n");
				fprintf(stderr, "There is NO WARRANTY, to the extent permitted by law.\n");
				exit(EXIT_SUCCESS);

			case 'd':
				device = optarg;
				break;

			case 'm':
				if (Parse_spi_mode(optarg, &new_config) != 0)
					exit(EXIT_FAILURE);
				break;

			case 'l':
				if (Parse_lsb_first(optarg, &new_config) != 0)
					exit(EXIT_FAILURE);
				break;

			case 's':
				if (Parse_spi_speed(optarg, &new_config) != 0)
					exit(EXIT_FAILURE);
				break;

			case 'r':
				if (Parse_spi_ready(optarg, &new_config) != 0)
					exit(EXIT_FAILURE);
				break;

			case 'B':
				if (Parse_spi_bits_per_word(optarg, &new_config) != 0)
					exit(EXIT_FAILURE);
				break;

			case 'b':
				if ((sscanf(optarg, "%d", &block_size) != 1)
				 || (block_size <= 0)) {
					fprintf(stderr, "%s: wrong blocksize\n", argv[0]);
					exit(EXIT_FAILURE);
				}
				break;

			case 'n':
				if ((sscanf(optarg, "%d", &blocks_count) != 1)
				 || (blocks_count < -1)) {
					fprintf(stderr, "%s: wrong block number\n", argv[0]);
					exit(EXIT_FAILURE);
				}
				break;

			default:
				fprintf(stderr, "%s: wrong option. Use -h for help.\n", argv[0]);
				exit(EXIT_FAILURE);
		}
	}

	if (((rx_buffer = malloc(block_size)) == NULL)
	 || ((tx_buffer = malloc(block_size)) == NULL)) {
		fprintf(stderr, "%s: not enough memory to allocate two %d bytes buffers\n",
		                argv[0], block_size);
		exit(EXIT_FAILURE);
	}

	memset(rx_buffer, 0, block_size);
	memset(tx_buffer, 0, block_size);

	if (device == NULL) {
		fprintf(stderr, "%s: no device specified (use option -h for help).\n", argv[0]);
		exit(EXIT_FAILURE);
	}

	fd = open(device, O_RDONLY);
	if (fd < 0) {
		perror(device);
		exit(EXIT_FAILURE);
	}

	if (Read_spi_configuration(fd, &prev_config) != 0)
		exit(EXIT_FAILURE);

	if (new_config.spi_mode == -1)
		new_config.spi_mode = prev_config.spi_mode;

	if (new_config.lsb_first == -1)
		new_config.lsb_first = prev_config.lsb_first;

	if (new_config.bits_per_word == -1)
		new_config.bits_per_word = prev_config.bits_per_word;

	if (new_config.spi_speed == -1)
		new_config.spi_speed = prev_config.spi_speed;

	if (new_config.spi_ready == -1)
		new_config.spi_ready = prev_config.spi_ready;

	if (Write_spi_configuration(fd, &new_config) != 0)
		exit(EXIT_FAILURE);

	while ((blocks_count > 0) || (blocks_count == -1)) {
		for (offset = 0; offset < block_size; offset += nb) {
			nb = read(STDIN_FILENO, & (tx_buffer[offset]), block_size - offset);
			if (nb <= 0)
				break;
		}
		if ((nb <= 0) && (offset == 0))
			break;

		if (Transfer_spi_buffers(fd, tx_buffer, rx_buffer, offset) != 0) {
			perror("SPI_IOC_MESSAGE");
			break;
		}
		if (write(STDOUT_FILENO, rx_buffer, offset) <= 0)
			break;
		if (blocks_count > 0)
			blocks_count --;
	}

	free(rx_buffer);
	free(tx_buffer);

	if (Write_spi_configuration(fd, &prev_config) != 0)
		exit(EXIT_FAILURE);

	if (blocks_count != 0)
		return EXIT_FAILURE;

	return EXIT_SUCCESS;
}
